aten/src/ATen/native/cuda/IndexKernel.cu

```diff
@@ -0,0 +1,102 @@
+#include <ATen/native/Indexing.h>
+
+#include <ATen/ATen.h>
+#include <ATen/Dispatch.h>
+#include <ATen/native/TensorIterator.h>
+#include <ATen/native/cuda/Loops.cuh>
+#include <ATen/cuda/Array.h>
+
+namespace at { namespace native {
+
+template <int N>
+static OffsetCalculator<N> index_make_offset_calculator(const TensorIterator& iter) {
+  AT_ASSERT(N <= iter.ntensors());
+  std::array<const int64_t*, N> strides;
+  for (int i = 0; i < N; i++) {
+    strides[i] = iter.strides(i).data();
+  }
+  return OffsetCalculator<N>(iter.ndim(), iter.shape().data(), strides.data());
+}
+
+template <typename func_t>
+void gpu_index_kernel(TensorIterator& iter, IntList index_size, IntList index_stride, const func_t& f) {
+  int num_indices = index_size.size();
+  AT_ASSERT(num_indices == index_stride.size());
+  AT_ASSERT(num_indices == iter.ntensors() - 2);
+
+  if (iter.numel() == 0) {
+    return;
+  }
+
+  auto sizes = cuda::Array<int64_t, 25>(0);
```

Comment:
I seem to recall that cuda Tensors are not allowed to have more than 25 dimensions right? This is the reason for this hardcoded size?
I juste tried but nothing prevents me from creating a cuda tensor with 30 dimensions. Where should this limit in the number of dimensions be inforced?

Reply:
After coalescing dimensions the the tensor can't have more than 25 dimensions. In a contiguous tensor, all the dimensions can be coalesced.

We should probably enforce this in TensorIterator, which we're not doing yet.

(We were talking about a hard limit of 64 dimensions for any tensor, but we haven't added code to enforce that yet.)